README.md
A simple clean way to format intervals with Javascript

# Setup

```sh
npm install simple-format-timeago --save
```

# Usage

```js
import formatTimeago from 'simple-format-timeago';

// handle now
formatTimeago(new Date()); // => 'Just now'
// handle past
formatTimeago(60 * 1000); // => '1 minute ago'
// handle plurals
formatTimeago(2 * 60 * 1000); // => '2 minutes ago'
// handle future
formatTimeago(-2 * 60 * 1000); // => '2 minutes from now'
```

Changing i18n

```js
var it = { // italian
  labels: {
    now: 'Adesso',
    seconds: 'secondi',
    second: 'secondo',
    minutes: 'minuti',
    minute: 'minuto',
    hours: 'ore',
    hour: 'ora',
    days: 'giorni',
    day: 'giorno',
    weeks: 'settimane',
    week: 'settimana',
    months: 'mesi',
    month: 'mese',
    years: 'anni',
    year: 'anno'
  },
  templates: {
    past: '<%= count %> <%= unit %> fa',
    future: 'tra <%= count %> <%= unit %>'
  }
};

formatTimeago(60 * SECOND, it); // => '1 minuto fa'
```

# API

```js
formatTimeago(date: Date | number, options: Object) => string
```

where:

- `date` a date (in the past or the future) or an interval expressed in milliseconds
- `options`
  - `labels`
    - `now` (default `'Just now'`)
    - `seconds` (default `'seconds'`)
    - `second` (default `'second'`)
    - `minutes` (default `'minutes'`)
    - `minute` (default `'minute'`)
    - `hours` (default `'hours'`)
    - `hour` (default `'hour'`)
    - `days` (default `'days'`)
    - `day` (default `'day'`)
    - `weeks` (default `'weeks'`)
    - `week` (default `'week'`)
    - `months` (default `'months'`)
    - `month` (default `'month'`)
    - `years` (default `'years'`)
    - `year` (default `'year'
  - `templates`
    - `past` (default `'<%= count %> <%= unit %> ago'`)
    - `future` (default `'<%= count %> <%= unit %> from now'`)

The following `locals` will be passed in to templates:

```js
{
  count: number; // the current count expressed in unit
  unit: string; // the current unit
  labels: {[key: string]: string;} // all labels
}
```

CHANGELOG.md
# Changelog

> **Tags:**
> - [New Feature]
> - [Bug Fix]
> - [Breaking Change]
> - [Documentation]
> - [Internal]
> - [Polish]

**Note**: Gaps between patch versions are faulty/broken releases.

## v0.0.1

First release


test/index.js
var test = require('tape');
var formatTimeago = require('..');
var SECOND = 1000;
var MINUTE = 60 * SECOND;
var HOUR = 60 * MINUTE;
var DAY = 24 * HOUR;
var WEEK = 7 * DAY;
var MONTH = 30 * DAY;
var YEAR = 12 * MONTH;

var it = {
  labels: {
    now: 'Adesso',
    seconds: 'secondi',
    second: 'secondo',
    minutes: 'minuti',
    minute: 'minuto',
    hours: 'ore',
    hour: 'ora',
    days: 'giorni',
    day: 'giorno',
    weeks: 'settimane',
    week: 'settimana',
    months: 'mesi',
    month: 'mese',
    years: 'anni',
    year: 'anno'
  },
  templates: {
    past: '<%= count %> <%= unit %> fa',
    future: 'tra <%= count %> <%= unit %>'
  }
};

test('should handle just now', function (assert) {
  assert.plan(2);
  assert.equal(formatTimeago(0), 'Just now');
  assert.equal(formatTimeago(999), 'Just now');
});

test('should handle minutes', function (assert) {
  assert.plan(4);

  // past
  assert.equal(formatTimeago(1 * SECOND), '1 second ago');
  assert.equal(formatTimeago(3 * SECOND), '3 seconds ago');

  // future
  assert.equal(formatTimeago(-1 * SECOND), '1 second from now');
  assert.equal(formatTimeago(-3 * SECOND), '3 seconds from now');
});

test('should handle minutes', function (assert) {
  assert.plan(12);

  // past
  assert.equal(formatTimeago(60 * SECOND), '1 minute ago');
  assert.equal(formatTimeago(61 * SECOND), '1 minute ago');
  assert.equal(formatTimeago(2 * 60 * SECOND - SECOND), '2 minutes ago');
  assert.equal(formatTimeago(2 * 60 * SECOND), '2 minutes ago');

  assert.equal(formatTimeago(60 * SECOND, it), '1 minuto fa');
  assert.equal(formatTimeago(2 * 60 * SECOND, it), '2 minuti fa');

  // future
  assert.equal(formatTimeago(-60 * SECOND), '1 minute from now');
  assert.equal(formatTimeago(-61 * SECOND), '1 minute from now');
  assert.equal(formatTimeago(-2 * 60 * SECOND + SECOND), '2 minutes from now');
  assert.equal(formatTimeago(-2 * 60 * SECOND), '2 minutes from now');

  assert.equal(formatTimeago(-60 * SECOND, it), 'tra 1 minuto');
  assert.equal(formatTimeago(-2 * 60 * SECOND, it), 'tra 2 minuti');
});

test('should handle hours', function (assert) {
  assert.plan(4);

  // past
  assert.equal(formatTimeago(1 * HOUR), '1 hour ago');
  assert.equal(formatTimeago(3 * HOUR), '3 hours ago');

  // future
  assert.equal(formatTimeago(-1 * HOUR), '1 hour from now');
  assert.equal(formatTimeago(-3 * HOUR), '3 hours from now');
});

test('should handle days', function (assert) {
  assert.plan(4);

  // past
  assert.equal(formatTimeago(1 * DAY), '1 day ago');
  assert.equal(formatTimeago(3 * DAY), '3 days ago');

  // future
  assert.equal(formatTimeago(-1 * DAY), '1 day from now');
  assert.equal(formatTimeago(-3 * DAY), '3 days from now');
});

test('should handle weeks', function (assert) {
  assert.plan(4);

  // past
  assert.equal(formatTimeago(1 * WEEK), '1 week ago');
  assert.equal(formatTimeago(3 * WEEK), '3 weeks ago');

  // future
  assert.equal(formatTimeago(-1 * WEEK), '1 week from now');
  assert.equal(formatTimeago(-3 * WEEK), '3 weeks from now');
});

test('should handle months', function (assert) {
  assert.plan(4);

  // past
  assert.equal(formatTimeago(1 * MONTH), '1 month ago');
  assert.equal(formatTimeago(3 * MONTH), '3 months ago');

  // future
  assert.equal(formatTimeago(-1 * MONTH), '1 month from now');
  assert.equal(formatTimeago(-3 * MONTH), '3 months from now');
});

test('should handle years', function (assert) {
  assert.plan(4);

  // past
  assert.equal(formatTimeago(1 * YEAR), '1 year ago');
  assert.equal(formatTimeago(3 * YEAR), '3 years ago');

  // future
  assert.equal(formatTimeago(-1 * YEAR), '1 year from now');
  assert.equal(formatTimeago(-3 * YEAR), '3 years from now');
});

test('should handle options.templates', function (assert) {
  assert.plan(1);
  assert.equal(formatTimeago(60 * SECOND, { templates: {
    past: '<%= labels.minutes %>: <%= count %>'
  } }), 'minutes: 1');
});




karma.conf.js
var webpack = require('webpack');
var path = require('path');

module.exports = function getConfig(config) {
  config.set({
    browserNoActivityTimeout: 30000,
    browsers: [process.env.CONTINUOUS_INTEGRATION ? 'Firefox' : 'Chrome'],
    singleRun: true,
    frameworks: ['tap'],
    files: [
      'test/index.js'
    ],
    preprocessors: {
      'test/index.js': ['webpack']
    },
    webpack: {
      node: {
        fs: 'empty'
      },
      module: {
        preLoaders: [
          {
            test: /\.js$/,
            include: path.resolve('index.js'),
            loader: 'isparta'
          }
        ]
      },
      plugins: [
        new webpack.DefinePlugin({
          'process.env.NODE_ENV': JSON.stringify('test')
        })
      ]
    },
    webpackMiddleware: {
      noInfo: true
    },
    reporters: ['dots', 'coverage'],
    coverageReporter: {
      type: 'text'
    }
  });
};


lib/formatInterval.js
var template = require('lodash/string/template');
var assign = require('lodash/object/assign');
var times = require('./times');
var templates = {};
var defaults = {
  labels: {
    now: 'Just now',
    seconds: 'seconds',
    second: 'second',
    minutes: 'minutes',
    minute: 'minute',
    hours: 'hours',
    hour: 'hour',
    days: 'days',
    day: 'day',
    weeks: 'weeks',
    week: 'week',
    months: 'months',
    month: 'month',
    years: 'years',
    year: 'year'
  },
  templates: {
    past: '<%= units %> <%= label %> ago',
    future: '<%= units %> <%= label %> from now'
  }
};

function formatInterval(delta, options) {
  options = assign({}, defaults, options);
  var labels = options.labels;
  var abs = Math.abs(delta);
  if (abs < times.SECOND) {
    return labels.now;
  }
  var units, label;
  if (abs < times.MINUTE) {
    units = Math.round(abs / times.SECOND);
    label = units > 1 ? labels.seconds : labels.second;
  } else if (abs < times.HOUR) {
    units = Math.round(abs / times.MINUTE);
    label = units > 1 ? labels.minutes : labels.minute;
  } else if (abs < times.DAY) {
    units = Math.round(abs / times.HOUR);
    label = units > 1 ? labels.hours : labels.hour;
  } else if (abs < times.WEEK) {
    units = Math.round(abs / times.DAY);
    label = units > 1 ? labels.days : labels.day;
  } else if (abs < times.MONTH) {
    units = Math.round(abs / times.WEEK);
    label = units > 1 ? labels.weeks : labels.week;
  } else if (abs < times.YEAR) {
    units = Math.round(abs / times.MONTH);
    label = units > 1 ? labels.months : labels.month;
  } else {
    units = Math.round(abs / times.YEAR);
    label = units > 1 ? labels.years : labels.year;
  }
  var formatter = delta >= 0 ? options.templates.past : options.templates.future;
  if (typeof formatter === 'string') {
    // compile the template just once
    formatter = ( templates[formatter] = templates[formatter] || template(formatter) );
  }
  return formatter({
    units: units,
    label: label,
    labels: labels
  });
}

module.exports = formatInterval;

index.js
var template = require('lodash/string/template');
var assign = require('lodash/object/assign');
var templates = {};
var defaults = {
  labels: {
    now: 'Just now',
    seconds: 'seconds',
    second: 'second',
    minutes: 'minutes',
    minute: 'minute',
    hours: 'hours',
    hour: 'hour',
    days: 'days',
    day: 'day',
    weeks: 'weeks',
    week: 'week',
    months: 'months',
    month: 'month',
    years: 'years',
    year: 'year'
  },
  templates: {
    past: '<%= count %> <%= unit %> ago',
    future: '<%= count %> <%= unit %> from now'
  }
};
var SECOND = 1000;
var MINUTE = 60 * SECOND;
var HOUR = 60 * MINUTE;
var DAY = 24 * HOUR;
var WEEK = 7 * DAY;
var MONTH = 30 * DAY;
var YEAR = 12 * MONTH;

function formatTimeago(date, options) {
  var delta = date instanceof Date ? new Date() - date.getTime() : date;
  options = assign({}, defaults, options);
  var labels = options.labels;
  var abs = Math.abs(delta);
  if (abs < SECOND) {
    return labels.now;
  }
  var count, unit;
  if (abs < MINUTE) {
    count = Math.round(abs / SECOND);
    unit = count > 1 ? labels.seconds : labels.second;
  } else if (abs < HOUR) {
    count = Math.round(abs / MINUTE);
    unit = count > 1 ? labels.minutes : labels.minute;
  } else if (abs < DAY) {
    count = Math.round(abs / HOUR);
    unit = count > 1 ? labels.hours : labels.hour;
  } else if (abs < WEEK) {
    count = Math.round(abs / DAY);
    unit = count > 1 ? labels.days : labels.day;
  } else if (abs < MONTH) {
    count = Math.round(abs / WEEK);
    unit = count > 1 ? labels.weeks : labels.week;
  } else if (abs < YEAR) {
    count = Math.round(abs / MONTH);
    unit = count > 1 ? labels.months : labels.month;
  } else {
    count = Math.round(abs / YEAR);
    unit = count > 1 ? labels.years : labels.year;
  }
  var formatter = delta >= 0 ? options.templates.past : options.templates.future;
  if (typeof formatter === 'string') {
    // compile the template just once
    formatter = ( templates[formatter] = templates[formatter] || template(formatter) );
  }
  return formatter({
    count: count,
    unit: unit,
    labels: labels
  });
}

module.exports = formatTimeago;